### ISSUE TITLE
Turning on full metrics will lose 40% performance and needs to be optimized

### ISSUE THREAD

**Issue by letian0805** [traefik/traefik#8118]
### Do you want to request a *feature* or report a *bug*?

Feature

### What did you expect to see?
Version: v2.4, contain routers metrics.
Metrics provider: prometheus
Metrics: entryPoints, services, routers
Trun off accessLog.
Case 1:
Turn off all metrics,  traefik can get 14000 QPS.
Case 2:
Turn on all metrics, can get only 9400 QPS.

Performance of metrics needs to be optimized.

**Comment by rtribotte** [comment]
Hello @letian0805,

Thanks for your interest in Traefik!

Which metric provider are you using?
Can you share something reproducible or the methodology you've used to reach your conclusions?

Without knowing exactly the context of the performance drop you are facing, it will be difficult to address your issue.
Nonetheless, it comes without saying that enabling metrics or access log, or any middleware, which remain as additional computations, the performances will drop accordingly to those additional computations.

**Comment by letian0805** [comment]
Hi @rtribotte 

I'm using prometheus.

Metrics configures of case 1:
```toml
#[accesslog]

[metrics]
    [metrics.prometheus]
        entryPoint = "metrics"
        addEntryPointsLabels = false
        addServicesLabels = false
        addRoutersLabels = false
```
case 2:
```toml
#[accesslog]

[metrics]
    [metrics.prometheus]
        entryPoint = "metrics"
        addEntryPointsLabels = true
        addServicesLabels = true
        addRoutersLabels = true
```

**Comment by letian0805** [comment]
accessLog false
addEntryPointsLabels = false
addServicesLabels = false
addRoutersLabels = false
total: 100000,  cost: 7.007286 s, requests/s: 14270.859992, success rate: 100%

accessLog false
addEntryPointsLabels = true
addServicesLabels = false
addRoutersLabels = false
total: 100000,  cost: 8.156821 s, requests/s: 12259.677966, success rate: 100%

accessLog false
addEntryPointsLabels = true
addServicesLabels = true
addRoutersLabels = false
total: 100000,  cost: 9.188745 s, requests/s: 10882.879146, success rate: 100%

accessLog false
addEntryPointsLabels = true
addServicesLabels = true
addRoutersLabels = true
total: 100000,  cost: 10.635392 s, requests/s: 9402.567991, success rate: 100%

**Comment by letian0805** [comment]
@jorge07 Ok. I'll try it. But i have tried this
```
accessLog false
addEntryPointsLabels = true
addServicesLabels = true
addRoutersLabels = false
total: 100000, cost: 9.188745 s, requests/s: 10882.879146, success rate: 100%
```

**Comment by jorge07** [comment]
Yeah, with this config you've full visibility and Service labels aren't required anymore. The performance penalty it's important that's right.

**Comment by letian0805** [comment]
From the test results, each item will increase the performance loss by at least 10%.

**Comment by gouki777** [comment]
There's a big performance drop.

**Comment by letian0805** [comment]
Any update?

**Comment by traefiker** [comment]
Closed by #9168.

**Comment by chuckyz** [comment]
I can confirm performance is vastly improved here.

Traefik 2.8 pre-PR
31470

Traefik 2.8 post-9168
82259

This is on a 64-core box with only Traefik running on it load-balancing to 3 fasthttp backends.
